### ISSUE TITLE
Therminology changes

### ISSUE THREAD

**Issue by dmitry-shechtman** [Knagis/CommonMark.NET#70]
So the Professor is back on CommonMark, which is great news.

Somewhat less great is his renaming of core terms in both the spec and the C implementation (I understand those are by popular demand -- I guess that's what happens when seeks community approval for a specification).

The following have changed so far:

1. *Header* --> *Heading*
2. *Horizontal Rule* --> *Thematic Break* (WTF?)

Should CommonMark.NET reflect those changes? I don't mind renaming everything, but wouldn't changing e.g. `HeaderLevel` to `HeadingLevel` break compatibility?

**Comment by Knagis** [comment]
I think that a second property that uses the same backing field plus obsoleting the previous property.

Heading sounds fine but the thematic break... I agree with your statement there :)

As for the struct for header - on cmark it contains heading level and heading source type (atx or setext) which is currently in the BlockTag enumeration for us. I don't know if that is worth it (a struct should not impact GC - at least I think so) - in that it seemingly does not add any value.

**Comment by dmitry-shechtman** [comment]
So `BlockTag` will have something like

    [Obsolete]
    AtxHeader,

    AtxHeading = AtxHeader,

etc. Not very pretty IMHO, but will do.

Re: struct I agree that it adds no value, but that would be consistent with `ListData` and `FencedCodeData` (the extensions branch has quite a few of those, including `DocumentData`, where I moved `ReferenceMap`). This could actually result reduce the memory footprint if stuff like `Heading.Level` is declared as `byte`.

**Comment by dmitry-shechtman** [comment]
I couldn't resist jumping on the opportunity with two renaming suggestions of my own:

1. [Weak emphasis](http://talk.commonmark.org/t/weak-vs-strong-emphasis/1943)
2. [Unordered list](http://talk.commonmark.org/t/ordered-vs-unordered-list/1945)

I updated the top post with these and turned it into a checklist.

**Comment by Knagis** [comment]
In case of the heading the struct would only have a single field - which is why it does not have its own struct yet.

As for the weak emphasis and unordered list - both sounds fine with me but lets hear what people on the forum says.

**Comment by dmitry-shechtman** [comment]
I guess the struct is a matter of personal taste.

I renamed both headers and rulers in the extensions branch (leaving `HeaderLevel` untouched for now). I figured changes like these should be applied there rather than to master.

BTW After having read through the topic _thematic break_ started to make more sense to me. I don't care that much for the specific nomenclature, but it is reasonable to call it something different than _horizontal rule_. As for an _emphasis_ alternative, get ready for _emphatic stress_ as a viable option :scream:

**Comment by dmitry-shechtman** [comment]
Unlike with _weak emphasis_, which might be changed as proposed, changed to something unthinkable, or remain unchanged, I feel that _unordered lists_ is the only correct term.

With that in mind, I went ahead and did a _backwards-compatible_ rename. The state of lists in the extensions branch is now as follows:

1. `ListType` and `ListData.ListType` are obsolete (`ListType.Bullet` stands unchanged).
2. `BlockTag.List` is obsoleted by `BlockTag.UnorderedList` and `BlockTag.OrderedList`.
3. `ListData.BulletChar` is obsoleted by `UnorderedListData.BulletCharacter`.
4. `ListData.Start` is obsoleted by `OrderedListData.Start`.
5. `ListData.Delimiter` and `ListDelimiter` are obsoleted by `OrderedListData.DelimiterCharacter`.
6. `UnorderedListData` and `OrderedListData` define a few additional properties in support of FancyLists.
7. A LegacyLists extension provides a full backwards compatibility mode, including Unicode bullets (which are soon to become a FancyLists feature).

**Comment by dmitry-shechtman** [comment]
I started documenting the API changes [here](https://github.com/AMDL/CommonMark.NET/wiki/API-Changes).

As you may see, I ultimately moved `HeaderLevel` into `Heading.Level`. I moved `DelimiterCharacter` into `Emphasis.DelimiterCharacter` (but didn't document it since that's not in 0.10.0).

I believe that similarly turning `Linkable` into a struct and removing the `TargetUrl` and `LiteralContent` backing fields wouldn't impact performance.

**Comment by Knagis** [comment]
Have you thought about any mechanisms how extensions like FancyLists would be supported from external assemblies that cannot modify Block and Inline classes?

**Comment by dmitry-shechtman** [comment]
Extensively :smile:

Lists are actually extensibility taken to the extreme. Adding a new list style is as easy as:

```C#
    public class FullWidthLowerAlphaLists : CommonMarkExtension
    {
        protected override IEnumerable<IBlockDelimiterHandler> InitializeBlockDelimiterHandlers(CommonMarkSettings settings)
        {
            var parameters = new OrderedListItemParameters('ａ', 'ｚ', listStyle: "fullwidth-lower-alpha");
            var delimiter = new ListItemDelimiterParameters('．');
            yield return new AlphaListItemHandler(settings, parameters, delimiter);
        }
    }
```

Numerous extensibility points are provided (a couple more added just today), so developing a custom extension should be a no-brainer.

There is still the problem of autodiscovery, i.e. how to make external extensions available in the command line. I'm considering MEF, but adding an external dependency might be an issue (especially with .NET < 4.0). It looks like that warrants separate packages (e.g. `CommonMark.NET.Extensibility` and `CommonMark.NET.Console`).

**Comment by dmitry-shechtman** [comment]
[CommonMark 0.23](http://talk.commonmark.org/t/ann-commonmark-0-23/1948) has been released.

May I suggest the same changes be applied before 0.11 (if you're planning on releasing one), i.e.

1. `HeaderLevel` obsoleted by `HeadingData` struct with `Level` property
1. `DelimiterCharacter` replaced by `EmphasisData` struct with `DelimiterCharacter` property
1. `Linkable`???

**Comment by Knagis** [comment]
Yes, I don't see a reason for not doing this. If you merge the changes needed for 0.11 in master I could push the release to nuget.

**Comment by dmitry-shechtman** [comment]
I'm currently working on disallowing space between link text and link label.

```
Test Name:	Example541
Test Outcome:	Failed
Result Message:	Assert.AreEqual failed. Expected:<<p><img src="/url" alt="foo" title="title" />
[]</p>>. Actual:<<p><img src="/url" alt="foo" title="title[]" />
[]</p>>.
Result StandardOutput:	
Example 541
Section: 541

CommonMark:

    ![foo]␣
    []
    
    [foo]:␣/url␣"title"

Expected:

    <p><img␣src="/url"␣alt="foo"␣title="title"␣/>
    []</p>

Actual:

    <p><img␣src="/url"␣alt="foo"␣title="title[]"␣/>
    []</p>
```

Any ideas?

**Comment by Knagis** [comment]
https://github.com/Knagis/CommonMark.NET/blob/master/CommonMark/Parser/InlineMethods.cs#L935

I suspect that it will be enough to remove the check for space and newline.

**Comment by dmitry-shechtman** [comment]
Maybe my branch diverged too much from knagis/master, but that doesn't seem to solve the issue with this specific example.

**Comment by Knagis** [comment]
I will take a look at the new tests in a few hours.

**Comment by dmitry-shechtman** [comment]
No need, solved in the 0.11 branch.

**Comment by dmitry-shechtman** [comment]
#71 ready for review.
